File: apps/queue/service/models/queue_item.go
// Copyright 2023-2026 Ant Investor Ltd
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package models

import (
	"fmt"
	"time"

	"github.com/pitabwire/frame/data"
)

// QueueItemStatus enumerates queue item statuses.
type QueueItemStatus string

const (
	ItemStatusWaiting   QueueItemStatus = "waiting"
	ItemStatusServing   QueueItemStatus = "serving"
	ItemStatusCompleted QueueItemStatus = "completed"
	ItemStatusCancelled QueueItemStatus = "cancelled"
	ItemStatusNoShow    QueueItemStatus = "no_show"
	ItemStatusExpired   QueueItemStatus = "expired"
)

// QueueItem represents an item in a queue.
type QueueItem struct { //nolint:recvcheck // TableName() must be value receiver for GORM
	data.BaseModel `gorm:"embedded"`

	QueueID      string          `gorm:"column:queue_id;not null"`
	Priority     int             `gorm:"column:priority;not null;default:1"`
	Status       QueueItemStatus `gorm:"column:status;not null;default:waiting"`
	TicketNo     string          `gorm:"column:ticket_no;not null"`
	Category     string          `gorm:"column:category"`
	CustomerID   string          `gorm:"column:customer_id"`
	Metadata     string          `gorm:"column:metadata;type:jsonb"`
	CounterID    string          `gorm:"column:counter_id"`
	ServedBy     string          `gorm:"column:served_by"`
	CalledAt     *time.Time      `gorm:"column:called_at"`
	ServiceStart *time.Time      `gorm:"column:service_start"`
	ServiceEnd   *time.Time      `gorm:"column:service_end"`
	JoinedAt     time.Time       `gorm:"column:joined_at;not null"`
}

// TableName returns the database table name.
func (QueueItem) TableName() string {
	return "queue_items"
}

// ValidItemTransitions defines allowed status transitions.
var validItemTransitions = map[QueueItemStatus][]QueueItemStatus{ //nolint:gochecknoglobals // transition map
	ItemStatusWaiting:   {ItemStatusServing, ItemStatusCancelled, ItemStatusNoShow, ItemStatusExpired},
	ItemStatusServing:   {ItemStatusCompleted, ItemStatusCancelled, ItemStatusNoShow, ItemStatusWaiting},
	ItemStatusNoShow:    {ItemStatusWaiting}, // re-queue
	ItemStatusCompleted: {},
	ItemStatusCancelled: {},
	ItemStatusExpired:   {},
}

// TransitionTo validates and performs a status transition.
func (q *QueueItem) TransitionTo(newStatus QueueItemStatus) error {
	allowed := validItemTransitions[q.Status]
	for _, s := range allowed {
		if s == newStatus {
			q.Status = newStatus
			return nil
		}
	}

	return fmt.Errorf("invalid transition from %s to %s", q.Status, newStatus)
}

File: ui_queuestore/lib/src/models/queue_item.dart
/// Status of a queue item.
enum QueueItemStatus {
  waiting,
  serving,
  completed,
  cancelled,
  noShow,
  expired;

  static QueueItemStatus fromString(String value) {
    return switch (value) {
      'no_show' => QueueItemStatus.noShow,
      _ => QueueItemStatus.values.firstWhere(
          (s) => s.name == value,
          orElse: () => QueueItemStatus.waiting,
        ),
    };
  }

  String toApiString() {
    return switch (this) {
      QueueItemStatus.noShow => 'no_show',
      _ => name,
    };
  }
}

/// An item in a queue (a person/ticket waiting for service).
class QueueItem {
  const QueueItem({
    required this.id,
    required this.queueId,
    this.priority = 1,
    this.status = QueueItemStatus.waiting,
    this.ticketNo = '',
    this.category = '',
    this.customerId = '',
    this.counterId = '',
    this.servedBy = '',
    this.joinedAt,
    this.calledAt,
    this.serviceStart,
    this.serviceEnd,
    this.metadata,
    this.createdAt,
  });

  final String id;
  final String queueId;
  final int priority;
  final QueueItemStatus status;
  final String ticketNo;
  final String category;
  final String customerId;
  final String counterId;
  final String servedBy;
  final DateTime? joinedAt;
  final DateTime? calledAt;
  final DateTime? serviceStart;
  final DateTime? serviceEnd;
  final Map<String, dynamic>? metadata;
  final DateTime? createdAt;

  factory QueueItem.fromJson(Map<String, dynamic> json) {
    return QueueItem(
      id: json['id'] as String? ?? '',
      queueId: json['queue_id'] as String? ?? '',
      priority: json['priority'] as int? ?? 1,
      status: QueueItemStatus.fromString(json['status'] as String? ?? ''),
      ticketNo: json['ticket_no'] as String? ?? '',
      category: json['category'] as String? ?? '',
      customerId: json['customer_id'] as String? ?? '',
      counterId: json['counter_id'] as String? ?? '',
      servedBy: json['served_by'] as String? ?? '',
      joinedAt: json['joined_at'] != null
          ? DateTime.tryParse(json['joined_at'] as String)
          : null,
      calledAt: json['called_at'] != null
          ? DateTime.tryParse(json['called_at'] as String)
          : null,
      serviceStart: json['service_start'] != null
          ? DateTime.tryParse(json['service_start'] as String)
          : null,
      serviceEnd: json['service_end'] != null
          ? DateTime.tryParse(json['service_end'] as String)
          : null,
      metadata: json['metadata'] as Map<String, dynamic>?,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (priority != 1) 'priority': priority,
      if (category.isNotEmpty) 'category': category,
      if (customerId.isNotEmpty) 'customer_id': customerId,
      if (ticketNo.isNotEmpty) 'ticket_no': ticketNo,
      if (metadata != null) 'metadata': metadata,
    };
  }
}
